README.md
# infimath.js [![Build Status](https://travis-ci.org/ETHproductions/infimath.svg?branch=master)](https://travis-ci.org/ETHproductions/infimath) [![Coverage Status](https://coveralls.io/repos/github/ETHproductions/infimath/badge.svg?branch=master)](https://coveralls.io/github/ETHproductions/infimath?branch=master)

WIP arbitrary-precision math library for JavaScript.

## How do I use it?

Browser:

    <script src='https://rawgit.com/ETHproductions/infimath/master/src/infimath.js'></script>
    
Usage:

    var myNumber1 = new BigNumber(0.1);   // Creates a new BigNumber
    var myNumber2 = BigNumber(0.2);       // Also creates a new BigNumber
    var myNumber3 = BigNumber.parse(0.3); // This too
    
    myNumber1.plus(myNumber2);            // Pushes an addition operation to the queue
    console.log(myNumber1);               // 0.1 (value hasn't changed yet)
    
    myNumber1.calculate();                // Empties the queue and performs each calculation
    console.log(myNumber1);               // 0.3
    console.log(myNumber1.eq(myNumber3)); // true
    
    // All of these are valid
    var numbers = [
        BigNumber(1),
        BigNumber(-1),
        BigNumber(0.1),
        BigNumber(1e-5),
        BigNumber("1"),
        BigNumber("-1.1"),
        BigNumber("2.4e5"),
        BigNumber("89,374,920.723_254_1"),
        BigNumber("89 374 920.723 254 1"),
        BigNumber([1,234,567]),
        BigNumber(BigNumber(1))
    ];
    
    console.log(
        BigNumber(1e+30)
        .plus(1e-30)
        .minus(1e-31)
        .times(2)
        .negate()
        .calculate()
        .toString()
        ); // "-2000000000000000000000000000000.0000000000000000000000000000018"

Move stuff here soon.


History.md
# 0.0.3 / 2016-08-13

- Added BigNumber.prototype.negate() - Pushes a negation operation to the queue. Does not change the value of `this`.
- Added BigNumber.prototype.compare/cmp(b[, opts]) - Returns BigNumber.compare(this, b, opts).
- Added BigNumber.prototype.less/lt(b[, opts]) - Returns `true` if `this` is less than `b`.
- Added BigNumber.prototype.greater/gt(b[, opts]) - Returns `true` if `this` is greater than `b`.
- Added BigNumber.prototype.equal/eq(b[, opts]) - Returns `true` if `this` is equal to `b`.
- BigNumber.compare now throws an error if one of the arguments is not a BigNumber.

# 0.0.2 / 2016-08-02

- Added BigNumber.prototype.plus(a[, b][, c...]) - Pushes multiplication operations to the queue. Does not change the value of `this`.
- Now integrated with [Travis CI](https://travis-ci.org/ETHproductions/infimath) and [Coveralls.io](https://coveralls.io/github/ETHproductions/infimath), using Mocha and Istanbul.
- Added more tests to bring the coverage up to 100%.
- Bug fix: BigNumber.prototype.calculate() does not empty the queue.

# 0.0.1 / 2016-08-01

- Added BigNumber function/object - can be called as `BigNumber(x)` or `new BigNumber(x)`.
  - Properties:
    - align(a, b) - Aligns the data of two BigNumbers by removing/adding zeroes to each side.
	- compare(a, b[, opts]) - Compares two BigNumbers, returning `-1` if `a` is less; `+1` if `a` is greater; `0` if they are equal. If `opts.aligned` is true, skips the alignment step. If `opts.ignoreSign` is true, ignores the sign of the numbers.
  - Prototype properties:
    - data - The meat of the number.
	- decs - The amount of decimal places in the number.
	- sign - The sign of the number (`-1` for negative, `0` for zero, `+1` for positive).
	- queue - The operations requested to be performed on the number. Emptied with calculate().
    - toString() - Returns a string representation of the number that the BigNumber represents.
    - plus(a[, b][, c...]) - Pushes addition operations to the queue. Does not change the value of `this`.
    - minus(a[, b][, c...]) - Pushes subtraction operations to the queue. Does not change the value of `this`.
    - calculate() - Empties the queue by performing all requested calculations and updating the value of `this`.


test/misc.js
var chai = require('chai');
var expect = chai.expect;
var BigNumber = require("../src/infimath.js");

describe('BigNumber.prototype.calculate()', function () {
	it('should accept a function as an argument', function () {
		var a = BigNumber(2),
			b = null,
			c = null;
		a.plus(2);
		b = a.calculate(function (x) {
			return c = x + "";
		}) + "";
		expect(b).to.equal("4");
		expect(c).to.equal("4");
	});
	
	it('should fail when the queue contains an invalid operation', function () {
		var a = BigNumber(2),
			b = false;
		a.queue.push(["error", BigNumber(2)]);
		try {
			a.calculate();
		} catch (e) {
			b = true;
		}
		expect(b).to.be.true;
	});
});

test/plus.js
var chai = require('chai');
var expect = chai.expect;
var BigNumber = require("../src/infimath.js");

var tests = [
	// Each test is run as BigNumber(a).plus(b) === {data: c, decs: d, sign: e}
	// [a, [b], c, d, e]
	
	// 0 + any
	[    "0", [    0], [   0],  0,  0],
	[    "0", [    1], [   1],  0,  1],
	[    "0", [   -1], [   1],  0, -1],
	[    "1", [    0], [   1],  0,  1],
	[   "-1", [    0], [   1],  0, -1],
	
	// pos + pos
	[    "1", [    1], [   2],  0,  1],
	[    "1", [  999], [   1], -1,  1],
	[    "1", [ 1000], [1, 1],  0,  1],
	[  "999", [    1], [   1], -1,  1],
	[ "1000", [    1], [1, 1],  0,  1],
	
	// pos + neg
	[    "1", [   -1], [   0],  0,  0],
	[    "1", [   -2], [   1],  0, -1],
	[    "1", [-1001], [   1], -1, -1],
	[    "1", [-1002], [1, 1],  0, -1],
	[    "1", [-1003], [2, 1],  0, -1],
	[    "1", [-1000001], [1], -2, -1],
	[    "1", [-1000002], [1, 0, 1],  0, -1],
	["1000001", [-1], [   1], -2,  1],
	["1000002", [-1], [1, 0, 1],  0,  1],
	
	// neg + pos
	[   "-1", [    1], [   0],  0,  0],
	[   "-1", [    2], [   1],  0,  1],
	[   "-1", [ 1001], [   1], -1,  1],
	[   "-1", [ 1002], [1, 1],  0,  1],
	[   "-1", [ 1003], [2, 1],  0,  1],
	[   "-1", [1000001], [1],  -2,  1],
	[   "-1", [1000002], [1, 0, 1],  0,  1],
	["-1000001", [1], [   1], -2, -1],
	["-1000002", [1], [1, 0, 1],  0, -1],
	
	// neg + neg
	[   "-1", [   -1], [   2],  0, -1],
	[   "-1", [ -999], [   1], -1, -1],
	[   "-1", [-1000], [1, 1],  0, -1],
	
	// dec + dec
	[    "1", [  1.1], [100, 2],  1,  1],
	[  "1.1", [    1], [100, 2],  1,  1],
	[  "1.1", [  1.1], [200, 2],  1,  1],
	[ "7.86", [  2.4], [260,10],  1,  1],
	[  "0.1", [999.9], [     1], -1,  1],
	["999.9", [  0.1], [     1], -1,  1],
];

tests.forEach(function(test) {
	describe('BigNumber("' + test[0] + '").plus(' + test[1].join(", ") + ').calculate()', function() {
		var a = BigNumber.prototype.plus.apply(BigNumber(test[0]), test[1]).calculate();
		it('should have data of [' + test[2] + ']', function() {
			expect(a.data + '').to.equal(test[2] + '');
		});
		it('should have decs of ' + test[3], function() {
			expect(a.decs).to.equal(test[3]);
		});
		it('should have sign of ' + test[4], function() {
			expect(a.sign).to.equal(test[4]);
		});
	});
});


test/align.js
var chai = require('chai');
var expect = chai.expect;
var BigNumber = require("../src/infimath.js");

var tests = [		
	// Positive
	{a: "0", b: "0", decs: 0, a_data: [0], a_sign: 0, b_data: [0], b_sign: 0},
	
	{a: "1000", b: "0", decs: -1, a_data: [1], a_sign: 1, b_data: [0], b_sign: 0},
	{a: "1000", b: "1", decs: 0, a_data: [0, 1], a_sign: 1, b_data: [1, 0], b_sign: 1},
	{a: "1000", b: "1000", decs: -1, a_data: [1], a_sign: 1, b_data: [1], b_sign: 1},
	
	{a: "-1000", b: "0", decs: -1, a_data: [1], a_sign: -1, b_data: [0], b_sign: 0},
	{a: "-1000", b: "1", decs: 0, a_data: [0, 1], a_sign: -1, b_data: [1, 0], b_sign: 1},
	{a: "-1000", b: "1000", decs: -1, a_data: [1], a_sign: -1, b_data: [1], b_sign: 1},
	
	{a: "1000", b: "-1", decs: 0, a_data: [0, 1], a_sign: 1, b_data: [1, 0], b_sign: -1},
	{a: "1000", b: "-1000", decs: -1, a_data: [1], a_sign: 1, b_data: [1], b_sign: -1},
	
	{a: "-1000", b: "-1", decs: 0, a_data: [0, 1], a_sign: -1, b_data: [1, 0], b_sign: -1},
	{a: "-1000", b: "-1000", decs: -1, a_data: [1], a_sign: -1, b_data: [1], b_sign: -1},
	
	{a: ".001", b: "0", decs: 1, a_data: [1], a_sign: 1, b_data: [0], b_sign: 0},
	{a: ".001", b: "1", decs: 1, a_data: [1, 0], a_sign: 1, b_data: [0, 1], b_sign: 1},
	{a: ".001", b: ".001", decs: 1, a_data: [1], a_sign: 1, b_data: [1], b_sign: 1},
	{a: ".001", b: "1000", decs: 1, a_data: [1, 0, 0], a_sign: 1, b_data: [0, 0, 1], b_sign: 1},
	
	{a: [q = BigNumber(1), q.data.push(0), q][2], b: [r = BigNumber(1), r.data.push(0), r][2], decs: 0, a_data: [1], a_sign: 1, b_data: [1], b_sign: 1},
];

tests.forEach(function(test) {
	var a = BigNumber(test.a),
		b = BigNumber(test.b);
	BigNumber.align(a, b);
	describe('BigNumber.align(BigNumber(' + test.a + '), BigNumber(' + test.b + '))', function () {
		it('a should have data of [' + test.a_data + ']', function() {
			expect(a.data + '').to.equal(test.a_data + '');
		});
		it('a should have decs of ' + test.decs, function() {
			expect(a.decs).to.equal(test.decs);
		});
		it('a should have sign of ' + test.a_sign, function() {
			expect(a.sign).to.equal(test.a_sign);
		});
		
		it('b should have data of [' + test.b_data + ']', function() {
			expect(b.data + '').to.equal(test.b_data + '');
		});
		it('b should have decs of ' + test.decs, function() {
			expect(b.decs).to.equal(test.decs);
		});
		it('b should have sign of ' + test.b_sign, function() {
			expect(b.sign).to.equal(test.b_sign);
		});
	});
});


src/infimath.js
/*jshint node: true */

/* Preface
 *   Quote by JSLint:
 *   "There is exactly one right way to write your code, and anyone who doesn't write this way is a complete and total moron."
 *   I HATE JSLINT.
 *   </rant>
 */

"use strict";

// Creates a new BigNumber object.
function BigNumber(x) {
	if(!(this instanceof BigNumber)) return new BigNumber(x);
	if(x === undefined) return this;
	return BigNumber.parse(x);
}

// Creates a new BigNumber object.
BigNumber.parse = function(x) {
	if (x instanceof BigNumber) return x;
	var orig = x;
	var result = new BigNumber();
	result.data = [];
	result.decs = 0;
	result.sign = 0;
	result.queue = [];

	if (typeof x !== "string") {
		x += ""; // Force input to be a string
	}

	// Remove commas and whitespace, e.g. "1,234,567" or "1.234 567 89".
	x = x.replace(/[,_\s]/g, "");

	// At this point, we check if the input is invalid, and throw a SyntaxError if it is.
	if (!/^[+-]?(\d+\.?\d*|\.\d+)(e[+-]?\d+)?$/.test(x))
		throw new SyntaxError("Invalid BigNumber: " + orig);

	// If x contains a non-zero digit:
	//   If the first char in x is "-", the sign is -1.
	//   Otherwise, the sign is +1.
	if (/^[^e]*[1-9]/i.test(x)) {
		if (x[0] === "-") {
			result.sign = -1;
			x = x.slice(1);
		} else {
			result.sign = +1;
		}
	}

	// Testing for scientific notation
	var ereg = /e([+-]?\d+)$/i;
	if (ereg.test(x)) {
		var decs = Number(x.match(ereg)[1]); // The +X or -X part
		x = x.replace(ereg, ""); // Remove the scientific part
		result.decs = -decs / 3 | 0; // The minimum number of decimals that will end up in the data
		decs %= 3;

		// Special cases where in the non-scientific part:
		// - the decimal point has to be manually shifted, and/or
		// - zeroes have to be added or removed.
		if (decs === +1) {
			   if (/\.(.)/.test(x))  x = x.replace(/\.(.)/, "$1.");
		  else                       x = x.replace(/\.?$/, "0");
		} else if (decs === +2) {
			   if (/\.(..)/.test(x)) x = x.replace(/\.(..)/, "$1.");
		  else if (/\.(.)/ .test(x)) x = x.replace(/\.(.)/, "$10.");
		  else                       x = x.replace(/\.?$/, "00");
		} else if (decs === -1) {
			   if (/(.)\./ .test(x)) x = x.replace(/(.)\./, ".$1");
		  else if (/\./    .test(x)) x = x.replace(/\./, ".0");
		  else if (/0$/    .test(x)) x = x.replace(/0$/, "");
		  else if (/(.)$/  .test(x)) x = x.replace(/(.)$/, ".$1");
		} else if (decs === -2) {
			   if (/(..)\./.test(x)) x = x.replace(/(..)\./, ".$1");
		  else if (/(.)\./ .test(x)) x = x.replace(/(.)\./, ".0$1");
		  else if (/\./    .test(x)) x = x.replace(/\./, ".00");
		  else if (/00$/   .test(x)) x = x.replace(/00$/, "");
		  else if (/(.)0$/ .test(x)) x = x.replace(/(.)0$/, ".$1");
		  else if (/(..)$/ .test(x)) x = x.replace(/(..)$/, ".$1");
		  else if (/(.)$/  .test(x)) x = x.replace(/(.)$/, ".0$1");
		}
	}
	
	// Add extra zeroes to the decimal part for proper parsing.
	if(/\./.test(x)) {
		x += "00";
	}

	// While there are at least three decimal digits, add the first three to the beginning of the data.
	var decreg = /\.(...)/;
	while (decreg.test(x)) {
		result.data.unshift(Number(x.match(decreg)[1]));
		result.decs++;
		x = x.replace(decreg, ".");
	}

	// Remove decimal point and extraneous zeroes.
	x = x.replace(/\..*/, "");

	// While there are digits left in x, add the last three to the end of data.
	while (x) {
		result.data.push(Number(x.slice(-3)));
		x = x.slice(0, -3);
	}

	// Remove all extraneous zeroes from the data.
	while (result.data.length > 1 && result.data[0] === 0) {
		result.data.splice(0, 1);
		result.decs--;
	}
	while (result.data.length > 1 && result.data.slice(-1)[0] === 0) {
		result.data.splice(-1, 1);
	}

	return result;
};

// Aligns the data of two BigNumbers by padding either side with zeroes.
BigNumber.align = function (a, b) {
	while (a.decs < b.decs) {
		a.data.unshift(0);
		a.decs++;
	}
	while (a.decs > b.decs) {
		b.data.unshift(0);
		b.decs++;
	}
	
	while (a.data.length < b.data.length) {
		a.data.push(0);
	}
	while (a.data.length > b.data.length) {
		b.data.push(0);
	}
	
	while (1 in a.data && 1 in b.data && a.data[0] === 0 && b.data[0] === 0) {
		a.data.shift();
		b.data.shift();
		a.decs--;
		b.decs--;
	}
	while (1 in a.data && 1 in b.data && a.data.slice(-1)[0] === 0 && b.data.slice(-1)[0] === 0) {
		a.data.pop();
		b.data.pop();
	}
};

// Compares two BigNumbers, returning -1 if the first is smaller, +1 if the first is larger, or 0 if they are equal
// If opts is {aligned: true, ignoreSign: true}, the function will skip the alignment step and ignore the signs of the numbers.
BigNumber.compare = BigNumber.cmp = function(a, b, opts) {
	if (!(a instanceof BigNumber) || !(b instanceof BigNumber)) {
		throw new TypeError("Cannot compare non-BigNumbers");
	}
	opts = opts || {};
	if (!opts.aligned) BigNumber.align(a, b);
	if (!opts.ignoreSign) {
		if (a.sign < b.sign) return -1;
		else if (a.sign > b.sign) return +1;
	}
	for (var i = a.data.length; --i >= 0; ) {
		if (a.data[i] < b.data[i]) {
			if (!opts.ignoreSign) {
				return -a.sign;
			} else {
				return -1;
			}
		} else if (a.data[i] > b.data[i]) {
			if (!opts.ignoreSign) {
				return +a.sign;
			} else {
				return +1;
			}
		}
	}
	return 0;
};

BigNumber.prototype.compare = BigNumber.prototype.cmp = function(b, opts) {
	return BigNumber.compare(this, b, opts);
};

// Returns `true` if `this` is less than the input.
BigNumber.prototype.less = BigNumber.prototype.lt = function(b, opts) {
	return BigNumber.compare(this, b, opts) === -1;
};

// Returns `true` if `this` is greater than the input.
BigNumber.prototype.greater = BigNumber.prototype.gt = function(b, opts) {
	return BigNumber.compare(this, b, opts) === +1;
};

// Returns `true` if `this` is equal to the input.
BigNumber.prototype.equal = BigNumber.prototype.eq = function(b, opts) {
	return BigNumber.compare(this, b, opts) ===  0;
};

// Converts the BigNumber to a string containing its nuerical data.
BigNumber.prototype.toString = function() {
	var decs = this.decs,
		str = "",
		arr = this.data.slice();

	// Add extra zeroes to the end and beginning of str, respectively.
	if (decs < 0) str = Array(1 - decs).join("000");
	while (arr.length <= decs) arr.push(0);

	// Combine arr into one string, including a decimal point.
	str = arr.reduce(function(x, y) {
		return ("00" + y).slice(-3) + (decs-- === 0 ? "." : "") + x;
	},str);

	// Trim extra zeroes and decimal point.
	str = str.replace(/^(?:0(?!\.))+|\.0*$|(\..+?)0+$/g, "$1");

	// Add a negative sign, if necessary.
	if (this.sign === -1) str = "-" + str;

	// Return str.
	return str;
};

// Pushes an addition operation to the queue.
BigNumber.prototype.plus = function () {
	var args = [].slice.call(arguments);
	for (var i in args) {
		this.queue.push(["plus", BigNumber(args[i])]);
	}
	return this;
};

// Pushes a subtraction operation to the queue.
BigNumber.prototype.minus = function () {
	var args = [].slice.call(arguments);
	for (var i in args) {
		var b = BigNumber(args[i]);
		b.sign = -b.sign;
		this.queue.push(["plus", b]);
	}
	return this;
};

// Pushes a multiplication operation to the queue.
BigNumber.prototype.times = function () {
	var args = [].slice.call(arguments);
	for (var i in args) {
		var b = BigNumber(args[i]);
		b.sign = b.sign;
		this.queue.push(["times", b]);
	}
	return this;
};

// Pushes a negation operation to the queue.
BigNumber.prototype.negate = function () {
	this.queue.push(["negate"]);
	return this;
};

// Empties the BigNumber's calculation queue.
BigNumber.prototype.calculate = function (x) {
	if (typeof x === "function") {
		return x(this.calculate());
	}
	var i, j, carry;
	for (i = 0; i < this.queue.length; i++) {
		var arr = this.queue[i];
		var item = arr[1];
		if (item) item.calculate();
		if (arr[0] === "plus") {
			if (item.sign === 0) {
				continue;
			} else if (this.sign === 0){
				this.data = item.data;
				this.decs = item.decs;
				this.sign = item.sign;
			} else if (this.sign === item.sign) {
				BigNumber.align(this, item);
				for (carry = 0, j = 0; j < item.data.length; j++) {
					this.data[j] += item.data[j] + carry;
					carry = +(this.data[j] >= 1e3);
					this.data[j] %= 1e3;
				}
				if (carry) this.data.push(1);
			} else {
				BigNumber.align(this, item);
				this.sign = this.sign * BigNumber.compare(this, item, {aligned:true, ignoreSign:true});
				for (carry = 0, j = 0; j < this.data.length; j++) {
					this.data[j] -= item.data[j] + carry;
					carry = +(this.data[j] < 0);
					this.data[j] += 1e3;
					this.data[j] %= 1e3;
				}
				if (carry) {
					var c = 0;
					this.data = this.data.map(function(x, y) { return (x -= c) > 0 ? (y ? 999 : 1e3) - x : (c = 1, 0); });
				}
			}
		} else if (arr[0] === "times") {
			this.decs += item.decs;
			this.sign *= item.sign;
			if (this.sign === 0 || item.sign === 0) {
				this.data = [0];
				this.decs = 0;
			} else if (this.data.length === 1 && this.data[0] === 1) {
				this.data = item.data;
			} else if (item.data.length === 1 && item.data[0] === 1) {
				// this.data doesn't need to change
			} else {
				var a = [];
				for (i = 0, j = 0; i < this.data.length + item.data.length; ++i) {
					a.push(0);
				}
				for (i = 0; i < this.data.length; ++i) {
					for (j = 0; j < item.data.length; ++j) {
						a[i + j] += this.data[i] * item.data[j];
					}
				}
				var l = a.length;
				a.push(0);
				for (i = 0; i < l; ++i) {
					a[i + 1] += a[i] / 1e3 | 0;
					a[i] %= 1e3;
				}
				this.data = a;
			}
		} else if (arr[0] === "negate") {
			this.sign *= -1;
		} else {
			throw new Error("Could not understand queued operation: " + arr[0]);
		}
	}
	
	this.queue = []; // Empty queue
	
	while (1 in this.data && this.data[0] === 0) {
		this.data.shift();
		this.decs--;
	}
	while (1 in this.data && this.data.slice(-1)[0] === 0) {
		this.data.pop();
	}
	return this;
};

/* Addendum
 *   I might have liked JSLint, if it was the kind of person who fixes your code for you
 *   instead of the person from https://xkcd.com/1513/.
 */

if (typeof module === "object") module.exports = BigNumber;
